### PR TITLE
L1 tests refactoring

### PR DESCRIPTION
**Pull request by ndobromirov**
### Overview
This pull request implements the described L1 testing changes in #43 for APCu, Static and SQLite drivers.

### Summary
There is a new L1CacheTest abstract base class. 
There are 3 successors for it, 1 for each driver (APCu, Static, SQLite).

There were some incompatibilities between L1 Implementations and tests:
 - Pool sharing between instances was tested only for SQLite. Whe ported to test against all L1s, the tests for the static implementation failed. This is fixed in the driver, so all are currently passing.
 - APCu had a comment that it clears the whole APCu data on flush. This is re-implemented with an iterator, so now it should be safe.

This is based on top of #73, as the changes here depend on:
 - Unified L1 creation interface.
 - Factory for all L1 drivers.

### Description
All the pure L1 tests are moved to the L1 testing hierarchy.
TODO: L2 and Integrations.

### CODE REVIEW

File: src/L1CacheFactory.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        return $l1CacheInstance;
+    }
+
+    /**
+     * Factory method for the L1 APCu driver.
+     *
+     * @param string $pool
+     * @return \LCache\APCuL1
+     */
+    protected function createApcu($pool)
```
Comment (davidstrauss) [inline src/L1CacheFactory.php]: Should we call this `createAPCu`?

File: src/L1CacheFactory.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        return new StaticL1($pool, new StateL1Static());
+    }
+
+    /**
+     * Factory method for the L1 static driver.
+     *
+     * @param string $pool
+     * @return \LCache\SQLiteL1
+     */
+    protected function createSqlite($pool)
```
Comment (davidstrauss) [inline src/L1CacheFactory.php]: Should we call this `createSQLite`?

File: src/StateL1APCu.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * Utility method to reduce code duplication.
+     *
+     * @param string $key
+     *   Key to store the evet counters in.
+     *
+     * @return bool
+     *   True on success, false otherwise.
+     */
+    private function _recordEvent($key)
```
Comment (davidstrauss) [inline src/StateL1APCu.php]: Is there a reason this has an underscore prefix?

File: src/StateL1APCu.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * {inheritdoc}
+     */
+    public function clear()
+    {
+        apcu_store($this->statusKeyHits, 0);
+        apcu_store($this->statusKeyMisses, 0);
+        // TODO: Decide on how to handle the last applied event state on clear?
```
Comment (davidstrauss) [inline src/StateL1APCu.php]: I think resetting statistics and resetting the last applied event ID should be separate. The latter should, generally, only get flushed when the entire local cache gets flushed.

File: src/StateL1Null.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+namespace LCache;
+
+/**
+ * Fake/Stub state manager class.
+ *
+ * @author ndobromirov
+ */
+class StateL1Null extends StateL1Static implements StateL1Interface
```
Comment (davidstrauss) [inline src/StateL1Null.php]: I'm concerned about how this splits the `PHP_INT_MAX` for `getLastAppliedEventID` away from `NullL1`. It allows incorrect behavior when someone uses this state implementation with a stateful L1. As a rule, it should be possible to combine any StateL1 with any L1 and have correct (but maybe not ideal) behavior.

Perhaps, we should just not have a `StateL1Null` and use `StateL1Static` with `NullL1` when testing.

**Review by davidstrauss — CHANGES_REQUESTED** [review]
This is getting very, very close.

File: src/L1CacheFactory.php (hunk 1 above)
Comment (ndobromirov) [reply src/L1CacheFactory.php]: Yes

File: src/L1CacheFactory.php (hunk 2 above)
Comment (ndobromirov) [reply src/L1CacheFactory.php]: Yes

File: src/StateL1APCu.php (hunk 3 above)
Comment (ndobromirov) [reply src/StateL1APCu.php]: It's a personal preference. I like to have all private stuff prefixed with underscore old habit left from development on [Yii framework](http://www.yiiframework.com/doc/guide/1.1/en/basics.convention#code). If it's not ok with the project's naming conventions it fairly easy to change. Though the code sniffer did not complained for that :).

File: src/StateL1APCu.php (hunk 4 above)
Comment (ndobromirov) [reply src/StateL1APCu.php]: So it's ok as it is. I'll just remove the comment.

File: src/StateL1Null.php (hunk 5 above)
Comment (ndobromirov) [reply src/StateL1Null.php]: I felt something was off when I was implementing this. The only state implementation that caused issues was the NULL one. I think that state should be ok to be the static one. Will change it so.

File: src/StateL1APCu.php (hunk 3 above)
Comment (ndobromirov) [reply src/StateL1APCu.php]: Done. Also changed the new static variable in the static cache instance.

File: src/StateL1Null.php (hunk 5 above)
Comment (ndobromirov) [reply src/StateL1Null.php]: State NULL is deleted. On all places the static one is used now. The PHP_INT_MAX was moved back to the NullL1.

**Comment by ndobromirov** [comment]
I think I've missed the NullL1 cache to have it's own L1 test class.

**Comment by davidstrauss** [comment]
> I think I've missed the NullL1 cache to have it's own L1 test class.

I'm not sure what you mean. Should I hold off on merging until you make some more changes?

**Review by davidstrauss — APPROVED** [review]

**Comment by ndobromirov** [comment]
That's it for L1s (for now).

I have started working on the L2 but it is going to be tricky there, as I am still getting to know the API and tests for it.

You can decide how to do it.
1. Merge as it is now and have L2 + integration added as separate PRs.
2. One huge PR (so this will get bigger).

**Comment by davidstrauss** [comment]
> Merge as it is now and have L2 + integration added as separate PRs.

I always prefer working incrementally.

**Pull request merged**